model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java

```diff
@@ -187,6 +187,7 @@ private List<ChangeSet> getLiquibaseUnrunChangeSets(Liquibase liquibase) throws
     private Liquibase getLiquibase(Class modelType, Connection connection, String defaultSchema) throws LiquibaseException {
         MapLiquibaseConnectionProvider liquibaseProvider = session.getProvider(MapLiquibaseConnectionProvider.class);
         String modelName = ModelEntityUtil.getModelName(modelType);
+        modelName = modelName.contains("authz") ? "authz" : modelName;
```

Comment:
As far as I understand all authz objects (`permission`, `policy`, `resource`, `resourceServer` and `scope`) will live inside one store. It might be useful to initialize objects in certain order to leverage fact that some object is already existing so that it's possible to create foreign keys.

Looking at it again it seems it'll need some re-design. The reason being is that we cache validated models within `JpaMapStorageProviderFactory.validatedModels` set and currents implementation will visit the validation 5 times. 

As of now I can see two possible ways how to solve it.

1. change  `JpaMapStorageProviderFactory.validatedModels` to `JpaMapStorageProviderFactory.validatedModelNames` where we'd cache names.
2. remove this line and also `jpa-authz-changelog-1.xml` (also probably relocate `jpa-authz-*-changelog-1xml`s) and update the order of MapStorage creation within `MapAuthorizationStoreFactory`:

```
        // order of creating stores matters
        MapStorage resourceServerStore = mapStorageProvider.getStorage(ResourceServer.class);
        MapStorage scopeStore = mapStorageProvider.getStorage(Scope.class);
        MapStorage resourceStore = mapStorageProvider.getStorage(Resource.class);
        MapStorage policyStore = mapStorageProvider.getStorage(Policy.class);
        MapStorage permissionTicketStore = mapStorageProvider.getStorage(PermissionTicket.class);
```

I'm currently in favor of 2. option, WDYT? Maybe another 3. option would be the best.

Reply:
As discussed, I'd only make it a bit more robust like this (please note that the `modelName == null` moved up to avoid a NPE). Avoiding ping the DB 5 times could be a future optimization.

```
        if (modelName == null) {
            throw new IllegalStateException("Cannot find changlelog for modelClass " + modelType.getName());
        }
        modelName = modelName.startsWith("authz-") ? "authz" : modelName;
```

Reply:
Thank you, added note about possible future optimization to `JpaMapStorageProviderFactory`